WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,262 @@
+
+    tilegen::CoordinatePacker<size_t> coord_packer1(1);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(2));
+
+    tilegen::CoordinatePacker<size_t> coord_packer2(2);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(2));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(3));
+
+    tilegen::CoordinatePacker<size_t> coord_packer10(10);
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(-1));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(10));
+    EXPECT_ANY_THROW(coord_packer0.addCoordinate(11));
+}
+TEST_F(TestCartesianPower, TestCoordinatePackerSize)
+{
+    tilegen::CoordinatePacker<size_t> coord_packer0(0);
+    EXPECT_EQ(0, coord_packer0.size());
+    tilegen::CoordinatePacker<size_t> coord_packer1(1);
+    EXPECT_EQ(1, coord_packer1.size());
+    tilegen::CoordinatePacker<size_t> coord_packer2(2);
+    EXPECT_EQ(2, coord_packer2.size());
+    tilegen::CoordinatePacker<size_t> coord_packer10(10);
+    EXPECT_EQ(10, coord_packer10.size());
+}
+
+TEST_F(TestCartesianPower, TestCoordinatePacker1)
+{
+    tilegen::CoordinatePacker<size_t> coord_packer(1);
+    EXPECT_EQ(0, coord_packer.packed());
+    coord_packer.addCoordinate(0);
+    EXPECT_EQ(0, coord_packer.packed());
+}
+
+
+TEST_F(TestCartesianPower, TestCoordinatePacker2)
+{
+    tilegen::CoordinatePacker<size_t> coord_packer(2);
+    EXPECT_EQ(0, coord_packer.packed());
+    coord_packer.addCoordinate(1);
+    EXPECT_EQ(1, coord_packer.packed());
+    coord_packer.addCoordinate(0);
+    EXPECT_EQ(2, coord_packer.packed());
+    coord_packer.addCoordinate(1);
+    EXPECT_EQ(5, coord_packer.packed());
+    coord_packer.addCoordinate(0);
+    EXPECT_EQ(10, coord_packer.packed());
+    coord_packer.addCoordinate(1);
+    EXPECT_EQ(21, coord_packer.packed());
+}
+
+TEST_F(TestCartesianPower, TestCoordinatePacker10)
+{
+    tilegen::CoordinatePacker<size_t> coord_packer(10);
+    EXPECT_EQ(0, coord_packer.packed());
+    coord_packer.addCoordinate(1);
+    EXPECT_EQ(1, coord_packer.packed());
+    coord_packer.addCoordinate(2);
+    EXPECT_EQ(12, coord_packer.packed());
+    coord_packer.addCoordinate(3);
+    EXPECT_EQ(123, coord_packer.packed());
+    coord_packer.addCoordinate(5);
+    EXPECT_EQ(1235, coord_packer.packed());
+    coord_packer.addCoordinate(0);
+    EXPECT_EQ(12350, coord_packer.packed());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerSize2D)
+{
+    std::pair<size_t, size_t> size{45, 45};
+    EXPECT_EQ(size, cp.size_2d(5));
+    size = {45, 15};
+    EXPECT_EQ(size, tilegen::CartesianPower<Iterator>(base, 3).size_2d(5));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCoordinates2D)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 15);
+    std::pair<size_t, size_t> coordinates{1, 6};
+    EXPECT_EQ(coordinates, it.coordinates_2d());
+}
```

Comment:
Now that coordinate packing is moved into a tested class, this test can be replaced with one that just checks uniqueness of the set of generated coordinate pairs.